qcodes/instrument_drivers/stahl/stahl.py

```diff
@@ -0,0 +1,220 @@
+
+        self.add_parameter(
+            "voltage",
+            get_cmd=f"{self.parent.identifier} U{self._channel_string}",
+            get_parser=self._stahl_get_parser("V"),
+            set_cmd=self._set_voltage,
+            unit="V",
+            vals=Numbers(
+                -self.parent.voltage_range,
+                self.parent.voltage_range
+            )
+        )
+
+        self.add_parameter(
+            "current",
+            get_cmd=f"{self.parent.identifier} I{self._channel_string}",
+            get_parser=self._stahl_get_parser("mA"),
+            unit="mA",
+        )
+
+        self.add_parameter(
+            "is_locked",
+            get_cmd=self._get_lock_status
+        )
+
+    @staticmethod
+    def _stahl_get_parser(unit: str) -> Callable:
+        """
+        Upon querying the voltage and current, the response from the instrument
+        is `+/-yy,yyy V` or `+/-yy,yyy mA'. We strip the unit and replace the
+        comma with a dot.
+
+        Args:
+            unit: The unit to strip
+
+        Return:
+            parser: Parse a response to a float
+        """
+        regex = f"([\+\-]\d+,\d+ ){unit}$"
+
+        def parser(response: str) -> float:
+            result, = re.search(regex, response).groups()
+            return float(result.replace(",", "."))
+
+        return parser
+
+    def _set_voltage(self, voltage: float) -> None:
+        """
+        Args:
+            voltage
+        """
+        # Normalize the voltage in the range 0 to 1, where 0 is maximum negative
+        # voltage and 1 is maximum positive voltage
+        voltage_normalized = np.interp(
+            voltage,
+            self.parent.voltage_range * np.array([-1, 1]),
+            [0, 1]
+        )
+
+        send_string = f"{self.parent.identifier} CH{self._channel_string} {voltage_normalized:.5f}"
+        response = self.ask(send_string)
+
+        if response != self._acknowledge_reply:
+            logger.warning(f"Command {send_string} did not produce an acknowledge reply")
+
+    def _get_lock_status(self) -> bool:
+        """
+        A lock occurs when an output is overloaded
+
+        Return:
+            lock_status: True when locked
+        """
+        send_string = f"{self.parent.identifier} LOCK"
+
+        response = self.parent.visa_handle.query_binary_values(
+            send_string,
+            datatype='B',
+            header_fmt="empty"
+        )
+
+        chnr = self._channel_number - 1
+        channel_group = chnr // 4
+        lock_code_group = response[channel_group]
+        return format(lock_code_group, "b")[chnr % 4 + 1] == "1"
+
+
+class Stahl(VisaInstrument):
+    """
+    Stahl driver.
+
+    Args:
+        name
+        address: A serial port address
+    """
+    def __init__(self, name: str, address: str):
+        super().__init__(name, address, terminator="\r")
+        self.visa_handle.baud_rate = 115200
+
+        instrument_info = self._parse_idn_string(
```

Comment:
The problem with that is that the IDN string returned by the instrument contains information like the voltage range and number of channels that I need elsewhere in the code, but are typically not part of the IDN message. If `get_idn` would basically be what is now `_parse_idn_string` then these class attributes would need to be set as a side-effect of calling `get_idn`, because the return arguments of `get_idn` is defined by the definition in the `VisaInstrument` base class. And I think setting class attributes through side-effects of calling some function is just a bad programming practice.